### PR TITLE
feat: Smithy Dart

### PR DESCRIPTION
**Pull request by dnys1**
Migrates Smithy Dart code to the Amplify Flutter repo

### CODE REVIEW

**Comment by dnys1** [comment]
I pushed a big JSON file (`endpoints.json`) which tacked on another 20k lines to the PR. I tried to avoid tracking it but it's needed for the build_verify step, which I think makes it worth having.

**Review by haverchuck — APPROVED** [review]
It would be nice to add some documentation that provides a general overview of how these packages work together - not a blocker though.

File: packages/smithy/smithy/lib/src/http/http_operation.dart [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+    };
+    final body = protocol.serialize(input, specifiedType: FullType(Input));
+    var host = uri.host;
+    if (!endpoint.isHostnameImmutable && request.hostPrefix != null) {
+      var prefix = request.hostPrefix!;
+      if (input is HasLabel) {
+        prefix = expandHostLabel(prefix, input.labelFor);
+      }
+      host = '$prefix$host';
+    }
```
Comment (Jordan-Nelson) [inline packages/smithy/smithy/lib/src/http/http_operation.dart:166]: There are a lot of variables in here that could be final with some refactoring. It makes it a little easier to read.

```suggestion
    final prefix = _createPrefix(
      endpoint: endpoint,
      request: request,
      input: input,
    );
    final host = '$prefix${uri.host}';
```

```dart
  String? _createPrefix({
    required Endpoint endpoint,
    required HttpRequest request,
    required Input input,
  }) {
    if (!endpoint.isHostnameImmutable && request.hostPrefix != null) {
      if (input is HasLabel) {
        return expandHostLabel(request.hostPrefix!, input.labelFor);
      }
      return request.hostPrefix;
    }
    return null;
  }
```

File: packages/smithy/smithy/lib/src/types/timestamp.dart [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+// millisecond precision. If a string is provided, it MUST be a valid RFC 3339
+// string with no UTC offset and optional fractional precision
+// (for example, 1985-04-12T23:20:50.52Z).
+
+import 'package:http_parser/http_parser.dart';
+import 'package:intl/intl.dart';
+import 'package:smithy/ast.dart';
+
+/// {@template aws.smithy.timestamp}
+///
```
Comment (Jordan-Nelson) [inline packages/smithy/smithy/lib/src/types/timestamp.dart:27]: Add a comment or remove

File: packages/smithy/smithy/lib/src/types/union.dart [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+import 'package:built_value/serializer.dart';
+import 'package:smithy/smithy.dart';
+
+/// An union representation in Smithy, i.e. a value which is one of a fixed set
+/// of types.
+///
+/// For example, the following Smithy model:
+///
+/// ```
+/// @enum("YES": {}, "NO": {})
```
Comment (Jordan-Nelson) [inline packages/smithy/smithy/lib/src/types/union.dart:25]: I think the example needs to be updated

File: packages/smithy/goldens/analysis_options.yaml [hunk 4]
```diff
@@ -0,0 +1 @@
+include: package:lints/recommended.yaml
```
Comment (Jordan-Nelson) [inline packages/smithy/goldens/analysis_options.yaml:1]: Do we want to use amplify lints? Could be done later as well.

File: packages/smithy/smithy_aws/test/http/aws_retryer_test.dart [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+              case Outcome.retryRequest:
+                expectation = completes;
+                break;
+              case Outcome.maxAttemptsExceeded:
+              case Outcome.retryQuotaExceeded:
+                expectation = throwsA(isA<_TransientSmithyException>());
+                break;
+            }
+
+            // TODO(dnys1): Try to get fake_async to work properly to avoid
```
Comment (Jordan-Nelson) [inline packages/smithy/smithy_aws/test/http/aws_retryer_test.dart:156]: Whats the unnecessary delay?

**Review by Jordan-Nelson — APPROVED** [review]
I left some comments, but I wouldn't consider any of them blocking. Nice work.

File: packages/smithy/smithy_aws/test/http/aws_retryer_test.dart (hunk 5 above)
Comment (dnys1) [reply packages/smithy/smithy_aws/test/http/aws_retryer_test.dart:156]: There's a `Future.delayed` inside the retryer

File: packages/smithy/goldens/analysis_options.yaml (hunk 4 above)
Comment (dnys1) [reply packages/smithy/goldens/analysis_options.yaml:1]: Nah, generated code should meet `lints` but it's not worth trying to get it cleaner than that.

File: packages/smithy/smithy/lib/src/types/timestamp.dart (hunk 2 above)
Comment (dnys1) [reply packages/smithy/smithy/lib/src/types/timestamp.dart:27]: Lol

**Pull request merged**